bndtools.m2e/src/bndtools/m2e/MavenBuildRepository.java

```diff
@@ -0,0 +1,246 @@
+
+            if (!bundleFile.exists()) {
+                MavenUtil.execJarMojo(mavenProject, new NullProgressMonitor());
+            }
+
+            if (bundleFile.exists()) {
+                try (Jar bundle = new Jar(bundleFile)) {
+                    if (bsn.equals(bundle.getBsn()) && version.compareTo(new Version(bundle.getVersion())) == 0) {
+                        returnVal = bundleFile;
+                    }
+                }
+            }
+        }
+
+        if (returnVal != null && returnVal.exists()) {
+            if (listeners != null && listeners.length > 0) {
+                for (DownloadListener listener : listeners) {
+                    try {
+                        listener.success(returnVal);
+                    } catch (Throwable t) {}
+                }
+            }
+        }
+
+        return returnVal;
+    }
+
+    private String getBsnFromMavenProject(MavenProject mavenProject) throws Exception {
+        final File outputDir = new File(mavenProject.getBuild().getOutputDirectory());
+
+        if (outputDir.exists()) {
+            try (Jar jar = new Jar(outputDir)) {
+                String bsn = jar.getBsn();
+
+                return bsn;
+            }
+        }
+
+        return null;
+    }
+
+    private void init() {
+        inited = true;
+
+        final IProgressMonitor monitor = new NullProgressMonitor();
+
+        for (IMavenProjectFacade projectFacade : mavenProjectRegistry.getProjects()) {
+            final IProject project = projectFacade.getProject();
+
+            try {
+                final MavenProject mavenProject = getMavenProject(projectFacade, monitor);
+
+                final String bsn = getBsnFromMavenProject(mavenProject);
+
+                if (bsn != null) {
+                    Entry<IMavenProjectFacade,MavenProject> entry = new AbstractMap.SimpleImmutableEntry<>(projectFacade, mavenProject);
+                    bsnMap.put(bsn, entry);
+                }
+            } catch (Exception e) {
+                logger.logError("Unable to get bundle symbolic name for " + project.getName(), e);
+            }
+        }
+
+        mavenProjectRegistry.addMavenProjectChangedListener(this);
+    }
+
+    private MavenProject getMavenProject(final IMavenProjectFacade projectFacade, final IProgressMonitor monitor) throws CoreException {
+        MavenProject mavenProject = projectFacade.getMavenProject();
+
+        if (mavenProject == null) {
+            mavenProject = projectFacade.getMavenProject(monitor);
+        }
+
+        return mavenProject;
+    }
+
+    @Override
+    public boolean canWrite() {
+        return false;
+    }
+
+    @Override
+    public List<String> list(String pattern) throws Exception {
+        return null;
+    }
+
+    @Override
+    public SortedSet<Version> versions(String bsn) throws Exception {
+        if (!inited) {
+            init();
+        }
+
+        List<Version> versions = new ArrayList<>();
```

Comment:
Why a list here? There is no loop so only a single version can be found. I know the result is a list, but internally only a single version is possible.